source/server/listener_manager_impl.cc

```diff
@@ -340,67 +342,82 @@ void ListenerImpl::addFilterChain(
     const std::string& transport_protocol, const std::vector<std::string>& application_protocols,
     const envoy::api::v2::listener::FilterChainMatch_ConnectionSourceType source_type,
+    const std::vector<std::string>& source_ips,
+    const Protobuf::RepeatedField<Protobuf::uint32>& source_ports,
     const Network::FilterChainSharedPtr& filter_chain) {
   if (destination_ports_map.find(destination_port) == destination_ports_map.end()) {
     destination_ports_map[destination_port] =
         std::make_pair<DestinationIPsMap, DestinationIPsTriePtr>(DestinationIPsMap{}, nullptr);
   }
   addFilterChainForDestinationIPs(destination_ports_map[destination_port].first, destination_ips,
                                   server_names, transport_protocol, application_protocols,
-                                  source_type, filter_chain);
+                                  source_type, source_ips, source_ports, filter_chain);
 }
 
 void ListenerImpl::addFilterChainForDestinationIPs(
     DestinationIPsMap& destination_ips_map, const std::vector<std::string>& destination_ips,
     const std::vector<std::string>& server_names, const std::string& transport_protocol,
     const std::vector<std::string>& application_protocols,
     const envoy::api::v2::listener::FilterChainMatch_ConnectionSourceType source_type,
+    const std::vector<std::string>& source_ips,
+    const Protobuf::RepeatedField<Protobuf::uint32>& source_ports,
     const Network::FilterChainSharedPtr& filter_chain) {
   if (destination_ips.empty()) {
     addFilterChainForServerNames(destination_ips_map[EMPTY_STRING], server_names,
-                                 transport_protocol, application_protocols, source_type,
-                                 filter_chain);
+                                 transport_protocol, application_protocols, source_type, source_ips,
+                                 source_ports, filter_chain);
   } else {
     for (const auto& destination_ip : destination_ips) {
       addFilterChainForServerNames(destination_ips_map[destination_ip], server_names,
                                    transport_protocol, application_protocols, source_type,
-                                   filter_chain);
+                                   source_ips, source_ports, filter_chain);
     }
   }
 }
 
 void ListenerImpl::addFilterChainForServerNames(
-    ServerNamesMap& server_names_map, const std::vector<std::string>& server_names,
+    ServerNamesMapSharedPtr& server_names_map_ptr, const std::vector<std::string>& server_names,
     const std::string& transport_protocol, const std::vector<std::string>& application_protocols,
     const envoy::api::v2::listener::FilterChainMatch_ConnectionSourceType source_type,
+    const std::vector<std::string>& source_ips,
+    const Protobuf::RepeatedField<Protobuf::uint32>& source_ports,
     const Network::FilterChainSharedPtr& filter_chain) {
+  if (server_names_map_ptr == nullptr) {
```

Comment:
This seems an unrelated change?

Reply:
It's not actually unrelated. To make a long story short, the old code was:
1) Doing a copy
2) More importantly by doing the copy, it made using unique_ptr impossible for nested structures. This was not a problem previously but is a problem now with deeper nesting.

I could probably split this out into a different change, but I would rather not as I don't think it would be a trivial split.

Reply:
Fair.